src/EditorFeatures/Test/Snippets/RoslynLSPSnippetConvertTests.cs

```diff
@@ -0,0 +1,260 @@
+            return TestAsync(markup, expectedLSPSnippet, cursorPosition, ImmutableArray.Create(new SnippetPlaceholder("test", placeholders)), textChange);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.RoslynLSPSnippetConverter)]
+        public Task TestExtendSnippetTextChangeBackwardsForPlaceholderForwardsForCaret()
+        {
+            var markup =
+@"{|placeholder:test|} [|if (true)
+{
+}|] $$";
+
+            var expectedLSPSnippet =
+@"${1:test} if (true)
+{
+} $0";
+            MarkupTestFile.GetPositionAndSpans(markup, out var outString, out var cursorPosition, out IDictionary<string, ImmutableArray<TextSpan>> dictionary);
+            var stringSpan = dictionary[""].First();
+            var textChange = new TextChange(new TextSpan(stringSpan.Start, 0), outString.Substring(stringSpan.Start, stringSpan.Length));
+            var placeholders = dictionary["placeholder"].Select(span => span.Start).ToImmutableArray();
+            return TestAsync(markup, expectedLSPSnippet, cursorPosition, ImmutableArray.Create(new SnippetPlaceholder("test", placeholders)), textChange);
+        }
+
+        #endregion
+
+        #region LSP Snippet generation tests
+
+        [Fact, Trait(Traits.Feature, Traits.Features.RoslynLSPSnippetConverter)]
+        public Task TestForLoopSnippet()
+        {
+            var markup =
+@"[|for (var {|placeholder1:i|} = 0; {|placeholder1:i|} < {|placeholder2:length|}; {|placeholder1:i|}++)
+{$$
+}|]";
+
+            var expectedLSPSnippet =
+@"for (var ${1:i} = 0; ${1:i} < ${2:length}; ${1:i}++)
+{$0
+}";
+            MarkupTestFile.GetPositionAndSpans(markup, out var outString, out var cursorPosition, out IDictionary<string, ImmutableArray<TextSpan>> dictionary);
+            var stringSpan = dictionary[""].First();
+            var textChange = new TextChange(new TextSpan(stringSpan.Start, 0), outString);
+            var placeholders1 = dictionary["placeholder1"].Select(span => span.Start).ToImmutableArray();
+            var placeholders2 = dictionary["placeholder2"].Select(span => span.Start).ToImmutableArray();
+            return TestAsync(markup, expectedLSPSnippet, cursorPosition, ImmutableArray.Create(
+                new SnippetPlaceholder("i", placeholders1),
+                new SnippetPlaceholder("length", placeholders2)), textChange);
+        }
+
+        #endregion
+
+        protected static TestWorkspace CreateWorkspaceFromCode(string code)
+         => TestWorkspace.CreateCSharp(code);
+
+        private static async Task TestAsync(string markup, string expectedLSPSnippet, int? cursorPosition, ImmutableArray<SnippetPlaceholder> placeholders, TextChange textChange)
+        {
+            using var workspace = CreateWorkspaceFromCode(markup);
+            var document = workspace.CurrentSolution.GetDocument(workspace.Documents.First().Id);
+
+            if (document is not null)
```

Comment:
when can this be null?  and won't that mean the test jsut passes instead of validating something?